[Bug] occasional crashes in v1.3.7: "panic: sync: WaitGroup is reused before previous Wait has returned"

InfluxDB: version 1.3.7 (from linux_armhf release tarball)
OS: Linux 4.1.41 on armv7l

I experienced these crashes several times, also for older versions. There are points written all the time, also deletes are performed regularly by cron jobs and scripts.

```
[I] 2017-11-13T08:58:56Z Snapshot for path /var/lib/influxdb/data/_internal/monitor/32 written in 1.303492817s engine=tsm1
[I] 2017-11-13T08:58:57Z beginning level 1 compaction of group 0, 2 TSM files engine=tsm1
[I] 2017-11-13T08:58:57Z compacting level 1 group (0) /var/lib/influxdb/data/_internal/monitor/32/000000010-000000001.tsm (#0) engine=tsm1
[I] 2017-11-13T08:58:57Z compacting level 1 group (0) /var/lib/influxdb/data/_internal/monitor/32/000000011-000000001.tsm (#1) engine=tsm1
[I] 2017-11-13T08:59:00Z compacted level 1 group (0) into /var/lib/influxdb/data/_internal/monitor/32/000000011-000000002.tsm.tmp (#0) engine=tsm1
[I] 2017-11-13T08:59:00Z compacted level 1 2 files into 1 files in 2.476008924s engine=tsm1
[I] 2017-11-13T09:01:34Z retention policy shard deletion check commencing service=retention
[I] 2017-11-13T09:31:34Z retention policy shard deletion check commencing service=retention
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 288 [running]:
sync.(*WaitGroup).Wait(0x114d2018)
	/usr/local/go/src/sync/waitgroup.go:133 +0x100
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).disableLevelCompactions(0x114d2000, 0x11228100)
	/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:272 +0x68
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).SetCompactionsEnabled(0x114d2000, 0x8eb200)
	/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:207 +0x60
github.com/influxdata/influxdb/tsdb.(*Shard).SetCompactionsEnabled(0x11228100, 0x10d62d00)
	/go/src/github.com/influxdata/influxdb/tsdb/shard.go:424 +0x48
github.com/influxdata/influxdb/tsdb.(*Store).monitorShards(0x10d561c0)
	/go/src/github.com/influxdata/influxdb/tsdb/store.go:1247 +0x1d8
created by github.com/influxdata/influxdb/tsdb.(*Store).Open
	/go/src/github.com/influxdata/influxdb/tsdb/store.go:150 +0x21c
[I] 2017-11-14T02:01:23Z InfluxDB starting, version 1.3.7, branch 1.3, commit 2d474a3089bcfce6b472779be9470a1f0ef3d5e4
[I] 2017-11-14T02:01:23Z Go version go1.8.4, GOMAXPROCS set to 2
```

I changed some settings to reduce the memory footprint, in case they do matter:
```
query-log-enabled = false
cache-max-memory-size = 16777216 # 16 MB instead of 1048576000
reporting-disabled = true
logging-enabled = false
cache-snapshot-memory-size = 8388608 # 8 MB instead of 26214400
max-concurrent-queries = 50
query-timeout = "60s"
log-queries-after = "30s"
max-select-series = 1024
log-enabled = false
max-connection-limit = 20
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@
 
 - [#9065](https://github.com/influxdata/influxdb/pull/9065): Refuse extra arguments to influx CLI
 - [#9058](https://github.com/influxdata/influxdb/issues/9058): Fix space required after regex operator. Thanks @stop-start!
+- [#9109](https://github.com/influxdata/influxdb/issues/9109): Fix: panic: sync: WaitGroup is reused before previous Wait has returned
 
 ## v1.4.2 [2017-11-15]
 

--- a/tsdb/engine/tsm1/engine.go
+++ b/tsdb/engine/tsm1/engine.go
@@ -282,17 +282,44 @@ func (e *Engine) disableLevelCompactions(wait bool) {
 		e.levelWorkers += 1
 	}
 
+	// Hold onto the current done channel so we can wait on it if necessary
+	waitCh := e.done
+
 	if old == 0 && e.done != nil {
+		// It's possible we have closed the done channel and released the lock and another
+		// goroutine has attempted to disable compactions.  We're current in the process of
+		// disabling them so check for this and wait until the original completes.
+		select {
+		case <-e.done:
+			e.mu.Unlock()
+			return
+		default:
+		}
+
 		// Prevent new compactions from starting
 		e.Compactor.DisableCompactions()
 
 		// Stop all background compaction goroutines
 		close(e.done)
-		e.done = nil
+		e.mu.Unlock()
+		e.wg.Wait()
 
+		// Signal that all goroutines have exited.
+		e.mu.Lock()
+		e.done = nil
+		e.mu.Unlock()
+		return
+	}
+	e.mu.Unlock()
+
+	// Compaction were already disabled.
+	if waitCh == nil {
+		return
 	}
 
-	e.mu.Unlock()
+	// We were not the first caller to disable compactions and they were in the process
+	// of being disabled.  Wait for them to complete before returning.
+	<-waitCh
 	e.wg.Wait()
 }
 
@@ -323,15 +350,34 @@ func (e *Engine) enableSnapshotCompactions() {
 
 func (e *Engine) disableSnapshotCompactions() {
 	e.mu.Lock()
-
+	var wait bool
 	if e.snapDone != nil {
+		// We may be in the process of stopping snapshots.  See if the channel
+		// was closed.
+		select {
+		case <-e.snapDone:
+			e.mu.Unlock()
+			return
+		default:
+		}
+
 		close(e.snapDone)
-		e.snapDone = nil
 		e.Compactor.DisableSnapshots()
+		wait = true
+
+	}
+	e.mu.Unlock()
+
+	// Wait for the snapshot goroutine to exit.
+	if wait {
+		e.snapWG.Wait()
 	}
 
+	// Signal that the goroutines are exit and everything is stopped by setting
+	// snapDone to nil.
+	e.mu.Lock()
+	e.snapDone = nil
 	e.mu.Unlock()
-	e.snapWG.Wait()
 
 	// If the cache is empty, free up its resources as well.
 	if e.Cache.Size() == 0 {

--- a/tsdb/engine/tsm1/engine_test.go
+++ b/tsdb/engine/tsm1/engine_test.go
@@ -816,6 +816,35 @@ func TestEngine_CreateCursor_Descending(t *testing.T) {
 	}
 }
 
+// This test ensures that "sync: WaitGroup is reused before previous Wait has returned" is
+// is not raised.
+func TestEngine_DisableEnableCompactions_Concurrent(t *testing.T) {
+	t.Parallel()
+
+	e := MustOpenDefaultEngine()
+	defer e.Close()
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		for i := 0; i < 1000; i++ {
+			e.SetCompactionsEnabled(true)
+			e.SetCompactionsEnabled(false)
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		for i := 0; i < 1000; i++ {
+			e.SetCompactionsEnabled(false)
+			e.SetCompactionsEnabled(true)
+		}
+	}()
+	wg.Wait()
+}
+
 func BenchmarkEngine_CreateIterator_Count_1K(b *testing.B) {
 	benchmarkEngineCreateIteratorCount(b, 1000)
 }

